host-tools/callisto-lib/src/lib.rs

```diff
@@ -0,0 +1,327 @@
+    }
+}
+
+fn reframing_state_from_ffi(val: &ReframingState) -> callisto::ReframingState {
+    match val {
+        ReframingState::Detect => callisto::ReframingState::Detect,
+        ReframingState::Done => callisto::ReframingState::Done,
+        ReframingState::Wait {
+            target_correction,
+            cur_wait_time,
+        } => callisto::ReframingState::Wait {
+            target_correction: *target_correction,
+            cur_wait_time: *cur_wait_time,
+        },
+    }
+}
+
+fn reframing_state_to_ffi(val: &callisto::ReframingState) -> ReframingState {
+    match val {
+        callisto::ReframingState::Detect => ReframingState::Detect,
+        callisto::ReframingState::Done => ReframingState::Done,
+        callisto::ReframingState::Wait {
+            target_correction,
+            cur_wait_time,
+        } => ReframingState::Wait {
+            target_correction: *target_correction,
+            cur_wait_time: *cur_wait_time,
+        },
+    }
+}
+
+fn control_state_from_ffi(st: &ControlSt) -> callisto::ControlSt {
+    callisto::ControlSt {
+        z_k: st.z_k,
+        b_k: speed_change_from_ffi(&st.b_k),
+        steady_state_target: st.steady_state_target,
+        rf_state: reframing_state_from_ffi(&st.rf_state),
+    }
+}
+
+fn control_state_to_ffi(st: &callisto::ControlSt, rs: &mut ControlSt) {
+    rs.z_k = st.z_k;
+    rs.b_k = speed_change_to_ffi(&st.b_k);
+    rs.steady_state_target = st.steady_state_target;
+    rs.rf_state = reframing_state_to_ffi(&st.rf_state);
+}
+
+fn control_config_from_ffi(cfg: &ControlConfig) -> callisto::ControlConfig {
+    callisto::ControlConfig {
+        reframing_enabled: cfg.reframing_enabled,
+        wait_time: cfg.wait_time,
+        target_count: cfg.target_count,
+    }
+}
+
+unsafe fn vsi_from_ptr<'a>(ptr: *const ()) -> &'a [callisto::StabilityIndication] {
+    let usize_ptr = ptr as *const usize;
+    let data_ptr = usize_ptr.offset(1) as *const callisto::StabilityIndication;
+    return std::slice::from_raw_parts(data_ptr, *usize_ptr);
+}
+
+unsafe fn data_counts_from_ptr<'a>(ptr: *const ()) -> &'a [isize] {
+    let usize_ptr = ptr as *const usize;
+    let data_ptr = usize_ptr.offset(1) as *const isize;
+    return std::slice::from_raw_parts(data_ptr, *usize_ptr);
+}
+
+/// Runs the callisto algorithm
+///
+/// # Safety
+///
+/// - `config_ptr` needs to point to a valid memory address holding a
+///   `ControlConfig`
+/// - `stability_checks_ptr` needs to point to a valid memory block holding
+///   a `VecS n StabilityIndication`
+/// - `data_counts_ptr` needs to point to a valid memory address holding
+///   a `VecS n (DataCountS m)
+/// - `control_state_ptr` needs to point to a valid memory address holding
+///   a `ControlSt`
+#[no_mangle]
+pub unsafe extern "C" fn __c_callisto_rust(
+    config_ptr: *const (),
+    availability_mask: u32,
+    stability_checks_ptr: *const (),
+    data_counts_ptr: *const (),
+    control_state_ptr: *mut (),
```

Comment:
The C ABI on Sys-V for pointers is the same regardless of types AFAIK, so you could just put the types here that you expect, that saves you some casts later and makes the signature a bit more useful.

Reply:
Mhm :thinking:, I'm not fully convinced yet ...

Then they don't match the pointer types of the foreign import in Haskell any more, which is kind of confusing. Personally, I would like to also conretize the types on the Haskell side, but they are defined in `bittide` and thus not available in `bittide-rust-ffi`. 

The whole idea of having `bittide-rust-ffi` separated is that this allows us to statically compile the Rust lib into the Haskell library providing a nice way of separation. However, since there is no Rust in `bittide-rust-ffi` anymore, we may alos move the foreign call directly to `bittide`. How do you think about that? 

Then we can provide concrete types at both ends, but need the custom `Setup.lhs` in `bittide` as well ...

Reply:
Devil's advocate in me says they already don't match because of `u32` vs `CUInt` :stuck_out_tongue:  But that's a valid consideration. In the end the casting is a no-op so it doesn't hurt, but also making the type signature a bit more useful would be nice!

Whether we should put all this into `bittide` and then keep the custom `Setup.lhs` I'm not sure. It would make sense I think since this package is really small and doesn't really do anything useful on its own, so probably putting it into `bittide` would make sense?

Reply:
> Devil's advocate in me says they already don't match because of u32 vs CUInt

But `CUInt` is just a `Word32` [wrapped by a newtype](https://hackage.haskell.org/package/base-4.18.0.0/docs/Foreign-C-Types.html#t:CUInt), so why is this a missmatch?

Reply:
I integrated `bittide-rust-ffi` directly into `bittide` now added type information to all the pointers.

Reply:
> But CUInt is just a Word32 [wrapped by a newtype](https://hackage.haskell.org/package/base-4.18.0.0/docs/Foreign-C-Types.html#t:CUInt), so why is this a missmatch?

In practice it's not a mismatch, it's just that `Word32` means "always 32 bits" while `CUInt` means "whatever C considers an unsigned integer on the current platform". For all the platforms we care about it is 32 bits but just semantically it's sliiiightly different. With `Word32` you say "give me 32 bits please" while `CUInt` is more "give me something that can hold an unsigned integer that's probably big enough for what I need it".

But this is not a blocker at all, it's fine. But since we do care about it being exactly 32 bits as it's a bitmask I personally would prefer `Word32`.

Reply:
That's a good point. I changed it to `Word32` then.